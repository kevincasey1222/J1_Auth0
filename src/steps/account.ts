import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../config';
import { getAcctWeblink } from '../util/getAcctWeblink';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';
export const ACCOUNT_ENTITY_TYPE = 'auth0_account';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const webLink = getAcctWeblink(instance.config.domain);
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: `Auth0 Account`,
          name: 'Auth0 Account',
        },
        assign: {
          _key: `auth0-account:${instance.id}`,
          _type: ACCOUNT_ENTITY_TYPE,
          _class: 'Account',
          name: 'Auth0 Account',
          displayName: 'Auth0 Account',
          webLink: webLink,
        },
      },
    }),
  );

  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'Auth0 Account',
        _type: ACCOUNT_ENTITY_TYPE,
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
