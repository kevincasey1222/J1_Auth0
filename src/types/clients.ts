// Client refers to an application that Auth0 supports
// For the class that we use to talk to the management API, see managementClient.ts
//schema per https://auth0.com/docs/api/management/v2/#!/Clients/get_clients

export interface ClientsResponseJwtConfiguration {
  lifetime_in_seconds?: number;
  secret_encoded?: boolean;
  scopes?: object;
  alg?: string; //Algorithm used to sign JWTs. Can be HS256 or RS256.
}

export interface ClientsResponseEncryptionKey {
  pub?: string;
  cert?: string;
  subject?: string;
}

export interface ClientsResponseRefreshToken {
  rotation_type?: string; //one of rotating, non-rotating
  expiration_type?: string; //one of expiring, non-expiring
  leeway?: number; //Period in seconds where the previous refresh token can be exchanged without triggering breach detection
  token_lifetime?: number; //in seconds
  infinite_token_lifetime?: boolean; //takes precedence over token_lifetime
  idle_token_lifetime?: number;
  infinite_idle_token_lifetime?: boolean;
}

export interface Auth0Client {
  client_id?: string;
  tenant?: string; //name of the tenant this client belongs to
  name?: string; //min 1 char, does not allow < or > in name
  description?: string; //max length: 140 characters
  global?: boolean; //Whether this is your global 'All Applications' client representing legacy tenant settings (true) or a regular client (false).
  client_secret?: string; //be sure to redact this is recordings and rawData
  app_type?: string; //Can be spa, native, non_interactive, or regular_web.
  logo_uri?: string;
  is_first_party?: boolean;
  oidc_conformant?: boolean;
  callbacks?: string[]; //Comma-separated list of URLs whitelisted for Auth0 to use as a callback to the client after authentication.
  allowed_origins?: string[];
  web_origins?: string[];
  client_aliases?: string[];
  allowed_clients?: string[];
  allowed_logout_urls?: string[];
  grant_types?: string[];
  jwt_configuration?: ClientsResponseJwtConfiguration; //probably should redact
  signing_keys?: object[]; //definitely redact from recordings and rawData
  encryption_key?: ClientsResponseEncryptionKey; //probably should redact
  sso?: boolean; //Applies only to SSO clients and determines whether Auth0 will handle Single Sign On (true) or whether the Identity Provider will (false).
  sso_disabled?: boolean; //defaults to true
  cross_origin_auth?: boolean; //whether client can make cross-origin auth requests
  cross_origin_loc?: string;
  custom_login_page_on?: boolean;
  custom_login_page?: string; //the content (HTML, CSS, JS) of any custom login page
  custom_login_page_preview?: string;
  form_template?: string; //used for WS-Federation
  addons?: object; //many subfields listed in schema, all of which provides more objects
  token_endpoint_auth_method?: string; //Can be none (public client without a client secret), client_secret_post (client uses HTTP POST parameters), or client_secret_basic (client uses HTTP Basic).
  client_metadata?: object; //unspecified object
  mobile?: object; //additional config for native mobile apps
  initiate_login_uri?: string; //must be https
  native_social_login?: object; //config for Native Social Login support
  refresh_token?: ClientsResponseRefreshToken; //should this be redacted? Could be valid for security analysis
  organization_usage?: string; //can be 'deny' (default), 'allow', or 'require'
  organization_require_behavior?: string; //Can be no_prompt (default) or pre_login_prompt.
}
