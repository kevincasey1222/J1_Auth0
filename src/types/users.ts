export interface Auth0UserIdentity {
  connection: string;
  user_id: string;
  provider: string;
  isSocial: boolean;
}

//schema per https://auth0.com/docs/api/management/v2/#!/Users/get_users
export interface Auth0User {
  user_id?: string;
  email?: string;
  email_verified?: boolean;
  username?: string;
  phone_number?: string;
  phone_verified?: boolean;
  created_at?: string;
  updated_at?: string;
  identities?: Auth0UserIdentity[];
  app_metadata?: object;
  user_metadata?: object;
  picture?: string;
  name?: string;
  nickname?: string;
  multifactor?: string[]; //List of multi-factor authentication providers with which this user has enrolled.
  last_ip?: string;
  last_login?: string;
  logins_count?: number;
  blocked?: boolean;
  given_name?: string;
  family_name?: string;
}
