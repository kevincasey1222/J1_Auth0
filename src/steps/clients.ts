import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';

export async function fetchClients({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateClients(async (client) => {
    //the following properties are all security related and should be redacted
    delete client.client_secret;
    delete client.jwt_configuration;
    delete client.signing_keys;
    delete client.encryption_key;
    //the following properties contain objects with unknown data
    //deleting for security purposes, though it's possible they might be wanted later
    delete client.addons;
    delete client.client_metadata;
    delete client.mobile;
    delete client.native_social_login;
    //see schema in types/clients.ts for more on what these fields mean
    const clientEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: client,
          assign: {
            _key: client.client_id,
            _type: 'auth0_client',
            _class: 'Application',
            name: client.name,
            displayName: client.name,
            webLink:
              accountEntity.webLink +
              'applications/' +
              client.client_id +
              '/settings',
            clientId: client.client_id,
            tenant: client.tenant,
            description: client.description,
            global: client.global,
            appType: client.app_type,
            logoUri: client.logo_uri,
            isFirstParty: client.is_first_party,
            oidcConformant: client.oidc_conformant,
            callbacks: client.callbacks,
            allowedOrigins: client.allowed_origins,
            webOrigins: client.web_origins,
            clientAliases: client.client_aliases,
            allowedClients: client.allowed_clients,
            allowedLogoutUrls: client.allowed_logout_urls,
            grantTypes: client.grant_types,
            sso: client.sso,
            ssoDisabled: client.sso_disabled,
            crossOriginAuth: client.cross_origin_auth,
            crossOriginLoc: client.cross_origin_loc,
            customLoginPageOn: client.custom_login_page_on,
            customLoginPage: client.custom_login_page,
            customLoginPagePreview: client.custom_login_page_preview,
            formTemplate: client.form_template,
            tokenEndpointAuthMethod: client.token_endpoint_auth_method,
            initiateLoginUri: client.initiate_login_uri,
            organizationUsage: client.organization_usage,
            organizationRequireBehavior: client.organization_require_behavior,
            //the following fields are arguably security information useful to attackers
            //but also useful for security posture analysis
            tokenExpirationType: client.refresh_token?.expiration_type,
            tokenTokenLifetime: client.refresh_token?.token_lifetime,
            tokenInfiniteTokenLifetime:
              client.refresh_token?.infinite_token_lifetime,
            tokenIdleTokenLifetime: client.refresh_token?.idle_token_lifetime,
            tokenInfiniteIdleTokenLifetime:
              client.refresh_token?.infinite_idle_token_lifetime,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: clientEntity,
      }),
    );
  });
}

export const clientSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-clients',
    name: 'Fetch Clients',
    entities: [
      {
        resourceName: 'Auth0 Client',
        _type: 'auth0_client',
        _class: 'Application',
      },
    ],
    relationships: [
      {
        _type: 'auth0_account_has_client',
        _class: RelationshipClass.HAS,
        sourceType: 'auth0_account',
        targetType: 'auth0_client',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchClients,
  },
];
