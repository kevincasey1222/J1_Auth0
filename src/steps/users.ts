import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../config';
import { DATA_ACCOUNT_ENTITY } from './account';

export async function fetchUsers({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateUsers(async (user) => {
    const userEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: user,
          assign: {
            _key: user.user_id,
            _type: 'auth0_user',
            _class: 'User',
            name: user.name,
            displayName: user.name,
            username: user.username || '',
            nickname: user.nickname,
            email: user.email,
            webLink: accountEntity.webLink + 'users', //userids are hashed in user-specific URLs
            userId: user.user_id,
            emailVerified: user.email_verified,
            phoneNumber: user.phone_number,
            phoneVerified: user.phone_verified,
            createdAt: user.created_at,
            updatedAt: user.updated_at,
            identities: JSON.stringify(user.identities), //array of objects
            appMetadata: JSON.stringify(user.app_metadata), //object
            userMetadata: JSON.stringify(user.user_metadata), //object
            picture: user.picture, //url
            multifactor: user.multifactor, //string[]
            lastIp: user.last_ip,
            lastLogin: user.last_login,
            loginsCount: user.logins_count,
            blocked: user.blocked,
            givenName: user.given_name,
            familyName: user.family_name,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: userEntity,
      }),
    );
  });
}

export const userSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-users',
    name: 'Fetch Users',
    entities: [
      {
        resourceName: 'User',
        _type: 'auth0_user',
        _class: 'User',
      },
    ],
    relationships: [
      {
        _type: 'auth0_account_has_user',
        _class: RelationshipClass.HAS,
        sourceType: 'auth0_account',
        targetType: 'auth0_user',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchUsers,
  },
];
