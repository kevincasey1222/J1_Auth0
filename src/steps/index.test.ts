import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../config';
import { fetchUsers } from './users';
import { fetchAccountDetails } from './account';
import { integrationConfig } from '../../test/config';
import { fetchClients } from './clients';

test('should collect data', async () => {
  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchUsers(context);
  await fetchClients(context);

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'auth0_account' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        webLink: { type: 'string', format: 'url' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'webLink'],
    },
  });

  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'auth0_user' },
        name: { type: 'string' },
        displayName: { type: 'string' },
        email: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name', 'displayName', 'email'],
    },
  });

  const appClients = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Application'),
  );
  expect(appClients.length).toBeGreaterThan(0);
  expect(appClients).toMatchGraphObjectSchema({
    _class: ['Application'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'auth0_client' },
        name: { type: 'string' },
        tenant: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });
});
