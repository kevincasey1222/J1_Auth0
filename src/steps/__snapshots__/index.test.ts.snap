// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "auth0-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "Auth0 Account",
            "name": "Auth0 Account",
          },
        },
      ],
      "_type": "auth0_account",
      "createdOn": undefined,
      "displayName": "Auth0 Account",
      "id": "Auth0 Account",
      "name": "Auth0 Account",
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "auth0|60997dbb9201390068ed3573",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-05-10T18:38:51.757Z",
            "email": "adam.williams@jupiterone.com",
            "email_verified": false,
            "identities": Array [
              Object {
                "connection": "Username-Password-Authentication",
                "isSocial": false,
                "provider": "auth0",
                "user_id": "60997dbb9201390068ed3573",
              },
            ],
            "name": "adam.williams@jupiterone.com",
            "nickname": "adam.williams",
            "picture": "https://s.gravatar.com/avatar/9ea4fa51e96ace12abc2f580a55d8dd7?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fad.png",
            "updated_at": "2021-05-10T18:38:51.757Z",
            "user_id": "auth0|60997dbb9201390068ed3573",
          },
        },
      ],
      "_type": "auth0_user",
      "appMetadata": undefined,
      "blocked": undefined,
      "createdAt": "2021-05-10T18:38:51.757Z",
      "createdOn": undefined,
      "displayName": "adam.williams@jupiterone.com",
      "email": "adam.williams@jupiterone.com",
      "emailVerified": false,
      "familyName": undefined,
      "givenName": undefined,
      "identities": "[{\\"connection\\":\\"Username-Password-Authentication\\",\\"user_id\\":\\"60997dbb9201390068ed3573\\",\\"provider\\":\\"auth0\\",\\"isSocial\\":false}]",
      "lastIp": undefined,
      "lastLogin": undefined,
      "loginsCount": undefined,
      "multifactor": undefined,
      "name": "adam.williams@jupiterone.com",
      "nickname": "adam.williams",
      "phoneNumber": undefined,
      "phoneVerified": undefined,
      "picture": "https://s.gravatar.com/avatar/9ea4fa51e96ace12abc2f580a55d8dd7?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fad.png",
      "updatedAt": "2021-05-10T18:38:51.757Z",
      "userId": "auth0|60997dbb9201390068ed3573",
      "userMetadata": undefined,
      "username": "",
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/users",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "auth0|6094191d9201390068ecb1da",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "created_at": "2021-05-06T16:28:13.644Z",
            "email": "thomaskevincasey74@gmail.com",
            "email_verified": false,
            "identities": Array [
              Object {
                "connection": "Username-Password-Authentication",
                "isSocial": false,
                "provider": "auth0",
                "user_id": "6094191d9201390068ecb1da",
              },
            ],
            "name": "thomaskevincasey74@gmail.com",
            "nickname": "thomaskevincasey74",
            "picture": "https://s.gravatar.com/avatar/94067c015850e0f500a37895e62dc45f?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fth.png",
            "updated_at": "2021-05-06T16:28:13.644Z",
            "user_id": "auth0|6094191d9201390068ecb1da",
          },
        },
      ],
      "_type": "auth0_user",
      "appMetadata": undefined,
      "blocked": undefined,
      "createdAt": "2021-05-06T16:28:13.644Z",
      "createdOn": undefined,
      "displayName": "thomaskevincasey74@gmail.com",
      "email": "thomaskevincasey74@gmail.com",
      "emailVerified": false,
      "familyName": undefined,
      "givenName": undefined,
      "identities": "[{\\"connection\\":\\"Username-Password-Authentication\\",\\"user_id\\":\\"6094191d9201390068ecb1da\\",\\"provider\\":\\"auth0\\",\\"isSocial\\":false}]",
      "lastIp": undefined,
      "lastLogin": undefined,
      "loginsCount": undefined,
      "multifactor": undefined,
      "name": "thomaskevincasey74@gmail.com",
      "nickname": "thomaskevincasey74",
      "phoneNumber": undefined,
      "phoneVerified": undefined,
      "picture": "https://s.gravatar.com/avatar/94067c015850e0f500a37895e62dc45f?s=480&r=pg&d=https%3A%2F%2Fcdn.auth0.com%2Favatars%2Fth.png",
      "updatedAt": "2021-05-06T16:28:13.644Z",
      "userId": "auth0|6094191d9201390068ecb1da",
      "userMetadata": undefined,
      "username": "",
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/users",
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "PulUF0aKgsMvkoJRsyicnqNWnPViL9ru",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "callback_url_template": false,
            "callbacks": Array [],
            "client_id": "PulUF0aKgsMvkoJRsyicnqNWnPViL9ru",
            "cross_origin_auth": false,
            "custom_login_page_on": true,
            "global": false,
            "grant_types": Array [
              "authorization_code",
              "implicit",
              "refresh_token",
              "client_credentials",
            ],
            "is_first_party": true,
            "is_token_endpoint_ip_header_trusted": false,
            "name": "Default App",
            "oidc_conformant": true,
            "refresh_token": Object {
              "expiration_type": "non-expiring",
              "idle_token_lifetime": 1296000,
              "infinite_idle_token_lifetime": true,
              "infinite_token_lifetime": true,
              "leeway": 0,
              "rotation_type": "non-rotating",
              "token_lifetime": 2592000,
            },
            "sso_disabled": false,
            "tenant": "dev-jupiterone",
          },
        },
      ],
      "_type": "auth0_client",
      "allowedClients": undefined,
      "allowedLogoutUrls": undefined,
      "allowedOrigins": undefined,
      "appType": undefined,
      "callbacks": Array [],
      "clientAliases": undefined,
      "clientId": "PulUF0aKgsMvkoJRsyicnqNWnPViL9ru",
      "createdOn": undefined,
      "crossOriginAuth": false,
      "crossOriginLoc": undefined,
      "customLoginPage": undefined,
      "customLoginPageOn": true,
      "customLoginPagePreview": undefined,
      "description": undefined,
      "displayName": "Default App",
      "formTemplate": undefined,
      "global": false,
      "grantTypes": Array [
        "authorization_code",
        "implicit",
        "refresh_token",
        "client_credentials",
      ],
      "initiateLoginUri": undefined,
      "isFirstParty": true,
      "logoUri": undefined,
      "name": "Default App",
      "oidcConformant": true,
      "organizationRequireBehavior": undefined,
      "organizationUsage": undefined,
      "sso": undefined,
      "ssoDisabled": false,
      "tenant": "dev-jupiterone",
      "tokenEndpointAuthMethod": undefined,
      "tokenExpirationType": "non-expiring",
      "tokenIdleTokenLifetime": 1296000,
      "tokenInfiniteIdleTokenLifetime": true,
      "tokenInfiniteTokenLifetime": true,
      "tokenTokenLifetime": 2592000,
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/applications/PulUF0aKgsMvkoJRsyicnqNWnPViL9ru/settings",
      "webOrigins": undefined,
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "yu8tHR4MEqzMWInatMKDKsnd9WplZ560",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "allowed_clients": Array [],
            "app_type": "non_interactive",
            "callback_url_template": false,
            "callbacks": Array [],
            "client_aliases": Array [],
            "client_id": "yu8tHR4MEqzMWInatMKDKsnd9WplZ560",
            "cross_origin_auth": false,
            "custom_login_page_on": true,
            "global": false,
            "grant_types": Array [
              "client_credentials",
              "authorization_code",
            ],
            "is_first_party": true,
            "is_token_endpoint_ip_header_trusted": false,
            "name": "Auth0 Management API (Test Application)",
            "oidc_conformant": true,
            "refresh_token": Object {
              "expiration_type": "non-expiring",
              "idle_token_lifetime": 2592000,
              "infinite_idle_token_lifetime": true,
              "infinite_token_lifetime": true,
              "leeway": 0,
              "rotation_type": "non-rotating",
              "token_lifetime": 31557600,
            },
            "sso_disabled": false,
            "tenant": "dev-jupiterone",
            "token_endpoint_auth_method": "client_secret_post",
          },
        },
      ],
      "_type": "auth0_client",
      "allowedClients": Array [],
      "allowedLogoutUrls": undefined,
      "allowedOrigins": undefined,
      "appType": "non_interactive",
      "callbacks": Array [],
      "clientAliases": Array [],
      "clientId": "yu8tHR4MEqzMWInatMKDKsnd9WplZ560",
      "createdOn": undefined,
      "crossOriginAuth": false,
      "crossOriginLoc": undefined,
      "customLoginPage": undefined,
      "customLoginPageOn": true,
      "customLoginPagePreview": undefined,
      "description": undefined,
      "displayName": "Auth0 Management API (Test Application)",
      "formTemplate": undefined,
      "global": false,
      "grantTypes": Array [
        "client_credentials",
        "authorization_code",
      ],
      "initiateLoginUri": undefined,
      "isFirstParty": true,
      "logoUri": undefined,
      "name": "Auth0 Management API (Test Application)",
      "oidcConformant": true,
      "organizationRequireBehavior": undefined,
      "organizationUsage": undefined,
      "sso": undefined,
      "ssoDisabled": false,
      "tenant": "dev-jupiterone",
      "tokenEndpointAuthMethod": "client_secret_post",
      "tokenExpirationType": "non-expiring",
      "tokenIdleTokenLifetime": 2592000,
      "tokenInfiniteIdleTokenLifetime": true,
      "tokenInfiniteTokenLifetime": true,
      "tokenTokenLifetime": 31557600,
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/applications/yu8tHR4MEqzMWInatMKDKsnd9WplZ560/settings",
      "webOrigins": undefined,
    },
    Object {
      "_class": Array [
        "Application",
      ],
      "_key": "yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "callbacks": Array [],
            "client_id": "yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa",
            "global": true,
            "is_first_party": true,
            "name": "All Applications",
            "owners": Array [
              "mr|auth0|609315f6b9c14800692fd98f",
            ],
            "refresh_token": Object {
              "expiration_type": "non-expiring",
              "idle_token_lifetime": 1296000,
              "infinite_idle_token_lifetime": true,
              "infinite_token_lifetime": true,
              "leeway": 0,
              "rotation_type": "non-rotating",
              "token_lifetime": 2592000,
            },
            "tenant": "dev-jupiterone",
          },
        },
      ],
      "_type": "auth0_client",
      "allowedClients": undefined,
      "allowedLogoutUrls": undefined,
      "allowedOrigins": undefined,
      "appType": undefined,
      "callbacks": Array [],
      "clientAliases": undefined,
      "clientId": "yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa",
      "createdOn": undefined,
      "crossOriginAuth": undefined,
      "crossOriginLoc": undefined,
      "customLoginPage": undefined,
      "customLoginPageOn": undefined,
      "customLoginPagePreview": undefined,
      "description": undefined,
      "displayName": "All Applications",
      "formTemplate": undefined,
      "global": true,
      "grantTypes": undefined,
      "initiateLoginUri": undefined,
      "isFirstParty": true,
      "logoUri": undefined,
      "name": "All Applications",
      "oidcConformant": undefined,
      "organizationRequireBehavior": undefined,
      "organizationUsage": undefined,
      "sso": undefined,
      "ssoDisabled": undefined,
      "tenant": "dev-jupiterone",
      "tokenEndpointAuthMethod": undefined,
      "tokenExpirationType": "non-expiring",
      "tokenIdleTokenLifetime": 1296000,
      "tokenInfiniteIdleTokenLifetime": true,
      "tokenInfiniteTokenLifetime": true,
      "tokenTokenLifetime": 2592000,
      "webLink": "https://manage.auth0.com/dashboard/us/dev-jupiterone/applications/yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa/settings",
      "webOrigins": undefined,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "auth0-account:local-integration-instance",
      "_key": "auth0-account:local-integration-instance|has|auth0|60997dbb9201390068ed3573",
      "_toEntityKey": "auth0|60997dbb9201390068ed3573",
      "_type": "auth0_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "auth0-account:local-integration-instance",
      "_key": "auth0-account:local-integration-instance|has|auth0|6094191d9201390068ecb1da",
      "_toEntityKey": "auth0|6094191d9201390068ecb1da",
      "_type": "auth0_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "auth0-account:local-integration-instance",
      "_key": "auth0-account:local-integration-instance|has|PulUF0aKgsMvkoJRsyicnqNWnPViL9ru",
      "_toEntityKey": "PulUF0aKgsMvkoJRsyicnqNWnPViL9ru",
      "_type": "auth0_account_has_client",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "auth0-account:local-integration-instance",
      "_key": "auth0-account:local-integration-instance|has|yu8tHR4MEqzMWInatMKDKsnd9WplZ560",
      "_toEntityKey": "yu8tHR4MEqzMWInatMKDKsnd9WplZ560",
      "_type": "auth0_account_has_client",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "auth0-account:local-integration-instance",
      "_key": "auth0-account:local-integration-instance|has|yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa",
      "_toEntityKey": "yBj8wIc4Xb4AWUgg0tAluVcmvxjGovCa",
      "_type": "auth0_account_has_client",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "auth0_account",
    "auth0_user",
    "auth0_account_has_user",
    "auth0_client",
    "auth0_account_has_client",
  ],
  "numCollectedEntities": 6,
  "numCollectedRelationships": 5,
}
`;
